### latent-space-sputum
An assessment of transforming complex P.a. gene expression into autoencoder latent space.


### Background

  Our lab has collected RNAseq data from *P. aeruginosa* treated with sub-inhibitory concentrations of the common fermentation product ethanol, *P.a.* - *C. albicans* co-culture and *P.a.* grown in *ex vivo* CF sputum samples. While these three datasets were not from the same experiment, they may have some similarities due to the prevalence of both *P.a.* and *C.a.* as well as detectable ethanol in sputa, generally speaking. Moreover, the complexity of *P.a.* transcriptional responses in these datasets make them interesting test cases for the application of eADAGE latent feature space for analysis.

### Getting Started

  The project can be used to assess the performance of latent space models of *P.a.* gene expression on datasets of *P.a.* transcriptomics, requiring:
  * the latent space model and the gene expression dataset are of the same reference genome
  * biological groupings of the gene expression dataset are provided a priori


### TODO
- Provide vignette with example data
- Refine generated plots and exporting functionality


### Contact
Georgia.Doing.GR@Dartmouth.edu
